### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by MelonyLady**

### CODE REVIEW

File: homework-1/cash_register.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    def apply_discount(self):
+        if self.discount:
+            print(self.total_price * .90)
+            return self.total_price * .90
+        else:
+            print(self.total_price)
+            return self.total_price
+
+    def get_total(self):
+        self.total_price = sum(self.total_items[item] for item in self.total_items)
```
Comment (Ali-Parandeh) [inline homework-1/cash_register.py:41]: Excellent work here

File: homework-1/cash_register.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        # print(list(self.total_items))
+        print("Your basket has the following items:")
+        for key in self.total_items:
+            print(key)
+        return self.total_items
+
+    def reset_register(self):
+        self.total_items.clear()
+        self.total_price = 0
+        return print("next transaction please!")
```
Comment (Ali-Parandeh) [inline homework-1/cash_register.py:54]: You could add the register has been cleared - next transaction please to provide better logging

File: homework-4/sorted_matrix_search.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+def search_in_matrix(matrix, number_of_rows, target):
```
Comment (Ali-Parandeh) [inline homework-4/sorted_matrix_search.py:1]: This function is reliant on the user providing a rows in matrix number. What if the matrix is large and they cannot count the matrix size? We would want the function to still work when the matrix and target only is provided

File: homework-4/sorted_matrix_search.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    row_index = 0
+    column_index = number_of_rows - 1
```
Comment (Ali-Parandeh) [inline homework-4/sorted_matrix_search.py:3]: The variable naming makes reading the code hard here and thus reduce maintainability and understanding of this code. There is number of rows, then columns then rows and etc. and being assigned to each other

File: homework-4/sorted_matrix_search.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    while row_index < number_of_rows and column_index >= 0:
+        if matrix[row_index][column_index] == target:
+            print(f"Target found at {row_index, column_index}")
+            return True
+        elif matrix[row_index][column_index] > target:
+            column_index -= 1
+        elif matrix[row_index][column_index] < target:
+            row_index += 1
```
Comment (Ali-Parandeh) [inline homework-4/sorted_matrix_search.py:11]: Is the way we loop reliant on the values of inside the matrix itself? Are we assuming the matrix is sorted both row and column wise?

File: homework-3/generate_phrase.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+characters8 = "1234567"
+phrase8 = "7564321"
+
+generate_phrase(characters8, phrase8)
+
+# testing random characters in string but not outside since I am using.lower()
+characters9 = "!£$%@"
+phrase9 = "@$!%£"
+
+generate_phrase(characters9, phrase9)
```
Comment (Ali-Parandeh) [inline homework-3/generate_phrase.py:67]: Nice job testing these here with print statements. You could provide these as a test suite with unit tests

File: homework-3/generate_phrase.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+from collections import Counter
+
+
+def generate_phrase(characters, phrase):
+    if Counter(characters.lower()) >= Counter(phrase.lower()):
```
Comment (Ali-Parandeh) [inline homework-3/generate_phrase.py:5]: Excellent work here. Well done for using the counter class

File: homework-1/student.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    def remove_subject(self, subject):
+        self.subjects.pop(subject)
+        return print(f"'{subject}' and mark for '{subject}' removed")
+
+    def update_grade_for_subject(self, subject, mark):
+        self.subjects.update({subject: mark})
+        return print(f"'{subject}' mark has been updated to {mark}")
+
+    def view_all_subjects(self):
+        return print(self.subjects)
```
Comment (Ali-Parandeh) [inline homework-1/student.py:36]: Return not needed here. Print returns none anyway

File: homework-1/student.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    def overall_mark_average(self):
+        print("Average marks across courses:")
+        return print(mean(list(self.subjects.values())))
```
Comment (Ali-Parandeh) [inline homework-1/student.py:40]: Love this one. Well done

**Review by Ali-Parandeh — APPROVED** [review]
Excellent submissions. Well done for the hard work. Left several comments but mostly are suggestions to improve the code quality. 

Homework 1 grade: 90
Homework 3 grade: 100
Homework 4 grade: 85

File: homework-4/sorted_matrix_search.py (hunk 5 above)
Comment (MelonyLady) [reply homework-4/sorted_matrix_search.py:11]: Well, the question specified that they were sorted each way, so I went with it. But that is something to consider for the future.

**Comment by MelonyLady** [comment]
Thank you for your comments, they are good food for thought

**Pull request merged**